src/Aspire.Dashboard/Model/DashboardClient.cs

```diff
@@ -0,0 +1,355 @@
+/// </summary>
+/// <remarks>
+/// An instance of this type is created per service call, so this class should not hold onto any state
+/// expected to live longer than a single RPC request. In the case of streaming requests, the instance
+/// lives until the stream is closed.
+/// </remarks>
+internal sealed class DashboardClient : IDashboardClient
+{
+    private const string DashboardServiceUrlVariableName = "DOTNET_DASHBOARD_GRPC_ENDPOINT_URL";
+    private const string DashboardServiceUrlDefaultValue = "http://localhost:18999";
+
+    private readonly Dictionary<string, ResourceViewModel> _resourceByName = new(StringComparers.ResourceName);
+    private readonly CancellationTokenSource _cts = new();
+    private readonly TaskCompletionSource _whenConnected = new();
+    private readonly object _lock = new();
+
+    private readonly ILoggerFactory _loggerFactory;
+    private readonly ILogger<DashboardClient> _logger;
+
+    private ImmutableHashSet<Channel<ResourceViewModelChange>> _outgoingChannels = [];
+    private string? _applicationName;
+
+    private const int StateNone = 0;
+    private const int StateInitialized = 1;
+    private const int StateDisposed = 2;
+    private int _state;
+
+    private readonly GrpcChannel _channel;
+    private readonly DashboardService.DashboardServiceClient _client;
+
+    private Task? _connection;
+
+    public DashboardClient(ILoggerFactory loggerFactory)
+    {
+        _loggerFactory = loggerFactory;
+
+        _logger = loggerFactory.CreateLogger<DashboardClient>();
+
+        var address = GetAddressUri(DashboardServiceUrlVariableName, DashboardServiceUrlDefaultValue);
+
+        _logger.LogInformation("Dashboard configured to connect to: {address}", address);
+
+        // Create the gRPC channel. This channel performs automatic reconnects.
+        // We will dispose it when we are disposed.
+        _channel = CreateChannel();
+
+        _client = new DashboardService.DashboardServiceClient(_channel);
+
+        static Uri GetAddressUri(string variableName, string defaultValue)
+        {
+            try
+            {
+                var uri = Environment.GetEnvironmentVariable(variableName) ?? defaultValue;
+
+                return new Uri(uri);
+            }
+            catch (Exception ex)
+            {
+                throw new InvalidOperationException($"Error parsing URIs from environment variable '{variableName}'.", ex);
+            }
+        }
+
+        GrpcChannel CreateChannel()
+        {
+            var httpHandler = new SocketsHttpHandler
+            {
+                EnableMultipleHttp2Connections = true,
+                KeepAlivePingDelay = TimeSpan.FromSeconds(20),
+                KeepAlivePingTimeout = TimeSpan.FromSeconds(10),
+                KeepAlivePingPolicy = HttpKeepAlivePingPolicy.WithActiveRequests,
+                PooledConnectionIdleTimeout = TimeSpan.FromHours(2)
```

Comment:
What is the reason for the long lifetime? Almost every server will kill an idle connection much faster. It's probably better that the client kills it first to avoid a race between the client starting to use a connection and the server killing it before receiving a request.

Reply:
What value would you recommend?

Reply:
The default is 60 seconds. That's less than Kestrel's idle timeout (130 seconds). Not specifying anything and using the default is fine.

Whatever value we have is probably academic. The watch task is trying to always keep an active request on the connection, so we'll always have one long running connection that never times out.

Reply:
I removed this property setter, favouring the default.